clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java

```diff
@@ -724,6 +725,90 @@ public void testLargeAssignmentAndGroupWithNonEqualSubscription(boolean hasConsu
         assignor.assignPartitions(partitionsPerTopic, subscriptions);
     }
 
+    @Timeout(90)
+    @ParameterizedTest(name = TEST_NAME_WITH_CONSUMER_RACK)
+    @ValueSource(booleans = {false, true})
+    public void testAssignmentAndGroupWithNonEqualSubscriptionNotTimeout(boolean hasConsumerRack) {
+        initializeRacks(hasConsumerRack ? RackConfig.BROKER_AND_CONSUMER_RACK : RackConfig.NO_CONSUMER_RACK);
+        int topicCount = hasConsumerRack ? 50 : 500;
+        int partitionCount = 2_00;
+        int consumerCount = 2_0;
+
+        List<String> topics = new ArrayList<>();
+        Map<String, List<PartitionInfo>> partitionsPerTopic = new HashMap<>();
+        for (int i = 0; i < topicCount; i++) {
+            String topicName = getTopicName(i, topicCount);
+            topics.add(topicName);
+            partitionsPerTopic.put(topicName, partitionInfos(topicName, partitionCount));
+        }
+        for (int i = 0; i < consumerCount; i++) {
+            if (i % 4 == 0) {
+                subscriptions.put(getConsumerName(i, consumerCount),
+                        subscription(topics.subList(0, topicCount / 2), i));
+            } else {
+                subscriptions.put(getConsumerName(i, consumerCount),
+                        subscription(topics.subList(topicCount / 2, topicCount), i));
+            }
+        }
+
+        Map<String, List<TopicPartition>> assignment = assignor.assignPartitions(partitionsPerTopic, subscriptions);
+
+        for (int i = 1; i < consumerCount; i++) {
+            String consumer = getConsumerName(i, consumerCount);
+            if (i % 4 == 0) {
+                subscriptions.put(
+                        consumer,
+                        buildSubscriptionV2Above(topics.subList(0, topicCount / 2),
+                        assignment.get(consumer), generationId, i)
+                );
+            } else {
+                subscriptions.put(consumer,
+                        buildSubscriptionV2Above(topics.subList(topicCount / 2, topicCount),
+                        assignment.get(consumer), generationId, i)
+                );
+            }
+        }
+
+        assignor.assignPartitions(partitionsPerTopic, subscriptions);
+    }
+
+    @Test
+    public void testSubscriptionNotEqualAndAssignSamePartitionWith3Generation() {
```

Comment:
Thanks for the test! I wrote a similar test, `testEnsurePartitionsAssignedToHighestGeneration` but just the generation order was 0, 1, 2 instead of, per your suggestions, 0, 2, 1.  Do you think we could parametrize that test or reuse the same test style, to make things more consistent, but change the assignment order?  Also, is the title accurate? This happens in `allSubscriptionsEqual`, so it doesn't matter if the subscriptions are equal or not right? The goal of this is to ensure consumer with the highest generation gets the right assignment, regardless of the generation order.

Reply:
this method is to test ``allSubscriptionsEqual`` so its name is a mistake, thx!

but this bug in ``allSubscriptionsEqual`` not lead to the specified partition that previously assigned to at least 3 generation consumers cannot be owned by the highest generation, but would lead this partition assign to more than one  consumers so  add to ``assignedPartitions`` more than once and also the function ``getUnassignedPartitions`` will get more partitions than actuality, as a result, many partitions already assigned to consumers will be added to ``unassignedPartitions``, all of these partitions would be assigned twice.

I couldn't understand how to reuse such test style, could you provide any example?